File: typescript/2463.MinimumTotalDistanceTraveled.ts
/**
 * Finds the minimum total distance traveled by robots to factories.
 * @param {number[]} robot - The list of robot positions.
 * @param {number[][]} factory - The list of factory positions and limits.
 * @returns {number} - The minimum total distance traveled.
 */
function minimumTotalDistance(robot: number[], factory: number[][]): number {
    // Sort robots and factories
    robot.sort((a, b) => a - b);
    factory.sort((a, b) => a[0] - b[0]);
    
    // Create memoization map
    const memo = new Map<string, number>();
    
    function dp(robotIdx: number, factoryIdx: number, used: number): number {
        // Base cases
        if (robotIdx === robot.length) return 0;
        if (factoryIdx === factory.length) return 1e15;
        
        // Create key for memoization
        const key = `${robotIdx},${factoryIdx},${used}`;
        if (memo.has(key)) return memo.get(key)!;
        
        // Try not using current factory
        let best = dp(robotIdx, factoryIdx + 1, 0);
        
        // Try using current factory if capacity allows
        if (used < factory[factoryIdx][1]) {
            const cost = Math.abs(robot[robotIdx] - factory[factoryIdx][0]);
            best = Math.min(best, dp(robotIdx + 1, factoryIdx, used + 1) + cost);
        }
        
        memo.set(key, best);
        return best;
    }
    
    return dp(0, 0, 0);
}

File: python/2463.MinimumTotalDistanceTraveled.py
from typing import List
from functools import lru_cache


class Solution:
    def minimumTotalDistance(self, robot: List[int], factory: List[List[int]]) -> int:
        """
        Finds the minimum total distance traveled by robots to factories.

        Args:
          robot (List[int]): The list of robot positions.
          factory (List[List[int]]): The list of factory positions and limits.

        Returns:
          int: The minimum total distance traveled.
        """
        robot.sort()
        factory.sort(key=lambda x: x[0])

        @lru_cache(None)
        def dp(robot_idx: int, factory_idx: int, used: int) -> int:
            """
            Args:
              robot_idx (int): The index of the robot to assign.
              factory_idx (int): The index of the factory to assign to.
              used (int): The number of robots already assigned to the current factory.

            Returns:
              int: The minimum total distance traveled.
            """
            # If all robots are assigned, we're done
            if robot_idx == len(robot):
                return 0
            # If no more factories but robots remain, invalid
            if factory_idx == len(factory):
                return 10**15

            # Try not using current factory
            best = dp(robot_idx, factory_idx + 1, 0)

            # Try using current factory if capacity allows
            if used < factory[factory_idx][1]:
                cost = abs(robot[robot_idx] - factory[factory_idx][0])
                best = min(best, dp(robot_idx + 1, factory_idx, used + 1) + cost)

            return best

        return dp(0, 0, 0)
